KAFKA-12803: Support reassigning partitions when in KRaft mode

cmccabe opened the pull request:
Support the KIP-455 reassignment API when in KRaft mode. Reassignments
which merely rearrange partitions complete immediately. Those that only
remove a partition complete immediately if the ISR would be non-empty
after the specified removals. Reassignments that add one or more
partitions follow the KIP-455 pattern of adding all the adding replicas
to the replica set, and then waiting for the ISR to include all the new
partitions before completing. Changes to the partition sets are
accomplished via PartitionChangeRecord.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -1204,6 +1317,298 @@ void generateLeaderAndIsrUpdates(String context,
+                                     ReassignablePartition partition,
+                                     List<ApiMessageAndVersion> records) {
+        // Check that the requested partition assignment is valid.
+        validateManualPartitionAssignment(partition.replicas(), OptionalInt.empty());
```
Should we further verify that after the reassignment, all partitions for the same topic have the same number of replicas?

cmccabe replied:
I think we allow changing this now so that different partitions have different replica counts, so not allowing this for KRaft mode would be an incompatibility.

junrao replied:
Hmm, when did we allow different partitions in the same topic to have different replica counts? This will be weird for min.isr since it is set at the topic level.

cmccabe replied:
I think reassignment has always supported giving different partitions in the same topic different replica counts. Example:
```
$ ./bin/kafka-topics.sh --bootstrap-server localhost:9092 --describe
Topic: foo      TopicId: o3KToxKaTX2rB6d-VufJig PartitionCount: 2       ReplicationFactor: 2    Configs: segment.bytes=1073741824
        Topic: foo      Partition: 0    Leader: 0       Replicas: 0,1   Isr: 0,1
        Topic: foo      Partition: 1    Leader: 1       Replicas: 1,0   Isr: 1,0

$ cat /tmp/reassignment.json
{"partitions": [{"topic": "foo", "partition": 1, "replicas": [0]}], "version":1}

$ ./bin/kafka-reassign-partitions.sh --bootstrap-server localhost:9092 --reassignment-json-file /tmp/reassignment.json --execute

$ ./bin/kafka-topics.sh --bootstrap-server localhost:9092 --describe
Topic: foo      TopicId: o3KToxKaTX2rB6d-VufJig PartitionCount: 2       ReplicationFactor: 2    Configs: segment.bytes=1073741824
        Topic: foo      Partition: 0    Leader: 0       Replicas: 0,1   Isr: 0,1
        Topic: foo      Partition: 1    Leader: 0       Replicas: 0     Isr: 0
```
Should we file a JIRA to possibly fix this in a later version of Kafka? It does seem like it could create problems. But changing it would also break compatibility, of course.

junrao replied:
Thanks, Colin. It seems that we don't enforce the same replication factor within a topic now. So, we can keep the logic as it is and have a followup jira to improve this.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/RemovingAndAddingReplicas.java:
```diff
@@ -0,0 +1,214 @@
+     * be removed.
+     *
+     * If a removing replica was in position X in the original replica list, it will
+     * appear in the merged list following the appearance of X non-new replicas.
```
The logic seems a bit complicated and I am not sure how useful it is. For example, if current and target are (1, 3, 4) and (1, 2, 4), this method returns (1, 3, 2, 4). The old code calculates merged replicas simply as (target ++ originReplicas).distinct. So, it would be (1,2,4,3). Any benefit of the new way of merging replicas?

cmccabe replied:
I think this can be simplified. I will take a look in a bit.

cmccabe replied:
In the latest change, I have split this class into `PartitionReassignmentReplicas` and `PartitionReassignmentRevert`,  which are both a lot simpler than `RemovingAndAddingReplicas` was.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/RemovingAndAddingReplicas.java:
```diff
@@ -0,0 +1,214 @@
+     *
+     * @return                      The merged replica list.
+     */
+    List<Integer> calculateMergedReplicas(List<Integer> currentReplicas,
```
The old class ReplicaAssignment stores the current/target list as well. Then one doesn't need to provide current and target again to compute merged replicas, which eliminates inconsistency. Could we do the same here?

cmccabe replied:
The issue is that we don't always have current and target. For example, if we're cancelling a reassignment, we just have current.

cmccabe replied:
I removed this class -- see above.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -1204,6 +1317,298 @@ void generateLeaderAndIsrUpdates(String context,
+        records.add(new ApiMessageAndVersion(record, (short) 0));
+    }
+
+    void changePartitionReassignment(String topicName,
```
Could we add some comments to document the partition reassignment flow? For example, what triggers the new replicas to be added, what triggers the completion of the reassignment, how are old replicas deleted, etc.

cmccabe replied:
I added more JavaDoc to `changePartitionReassignment`

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -1204,6 +1317,298 @@ void generateLeaderAndIsrUpdates(String context,
+                        Replicas.toArray(newIsr),
+                        configurationControl.uncleanLeaderElectionEnabledForTopic(topicName),
+                        r -> clusterControl.unfenced(r));
+                    if (!electionWasClean(newLeader, Replicas.toArray(newIsr))) {
```
We have this logic duplicated in multiple places. It's easy to forget about this part after each leader election. Could we factor that into bestLeader() such that it returns the new leader and the new isr?

cmccabe replied:
Hmm, can we do this in a follow-on? I started working on it, but it got bigger than I wanted for this PR...

cmccabe replied:
I refactored this logic into `PartitionChangeBuilder`

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -711,56 +830,73 @@ BrokersToIsrs brokersToIsrs() {
-                    setLeaderEpoch(partition.leaderEpoch).
-                    setCurrentIsrVersion(partition.partitionEpoch + 1).
-                    setIsr(partitionData.newIsr()));
+                PartitionChangeRecord record = generateLeaderAndIsrUpdate(topic,
```
In the old logic, when the reassignment completes, even if the leader doesn't need to change, we bump up the leader epoch to prevent the removed replicas from being added to back to ISR unnecessarily. generateLeaderAndIsrUpdate() doesn't seem to do the same thing?

cmccabe replied:
Yes, let's bump up the leader epoch here. Should be fixed now.

junrao commented on metadata/src/main/resources/common/metadata/PartitionChangeRecord.json:
```diff
@@ -29,6 +29,15 @@
-      "about": "-1 if there is now no leader; -2 if the leader didn't change; the new leader otherwise." }
+      "about": "-1 if there is now no leader; -2 if the leader didn't change; the new leader otherwise." },
+    { "name": "Replicas", "type": "[]int32", "default": "null", "entityType": "brokerId",
+      "versions": "0+", "nullableVersions": "0+", "taggedVersions": "0+", "tag": 2,
```
Should versions start from 1+ for the new fields?

cmccabe replied:
We have decided to reset versions to 0 since we dropped compatibility with 2.8. We bumped up the frame version so that the incompatibility can be detected.

junrao commented on metadata/src/main/resources/common/metadata/PartitionRecord.json:
```diff
@@ -27,9 +27,9 @@
     { "name": "Isr", "type":  "[]int32", "versions":  "0+",
       "about": "The in-sync replicas of this partition" },
-    { "name": "RemovingReplicas", "type":  "[]int32", "versions":  "0+", "nullableVersions": "0+", "entityType": "brokerId",
+    { "name": "RemovingReplicas", "type":  "[]int32", "versions":  "0+", "entityType": "brokerId",
```
Do we need to bump up the version by removing nullable?

cmccabe replied:
This should be addressed by the recent changes in trunk

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -235,6 +256,41 @@ int preferredReplica() {
+         * @param newIsr    The new ISR.
+         * @return          True if the reassignment is complete.
+         */
+        boolean isrChangeCompletesReassignment(int[] newIsr) {
```
The old controller disables auto leader balancing when a partition reassignment is in progress. Should we do the same thing here?

cmccabe replied:
We haven't implemented auto leader balancing yet, so we're OK here...

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -343,12 +409,21 @@ public String toString() {
+        }
+        TopicControlInfo prevTopic = topics.put(record.topicId(),
             new TopicControlInfo(record.name(), snapshotRegistry, record.topicId()));
+        if (prevTopic != null) {
```
Hmm, in this case, the topicId is the same, by removing the topic, we are deleting the current topicId. Perhaps we should throw an IllegalStateException since this is not expected?

cmccabe replied:
I'm going to take this part out for now to focus on just the reassignment stuff

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -1204,6 +1317,298 @@ void generateLeaderAndIsrUpdates(String context,
+            if (topicId != null) {
+                TopicControlInfo topicInfo = topics.get(topicId);
+                if (topicInfo == null) {
+                    throw new RuntimeException("No topic entry found for " + topicId);
```
Should we set an error code in the response instead of throwing an exception?

cmccabe replied:
I'm going to take this part out for now to focus on just the reassignment stuff

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -1204,6 +1317,298 @@ void generateLeaderAndIsrUpdates(String context,
+        return response;
+    }
+
+    ListPartitionReassignmentsResponseData listAllPartitionReassignments() {
```
This method is almost the same as listPartitionReassignments(). Could we reuse the logic somehow?

cmccabe replied:
Yes, I have refactored these to have much more in common

junrao commented on metadata/src/main/resources/common/metadata/PartitionChangeRecord.json:
```diff
@@ -29,6 +29,15 @@
+    { "name": "Replicas", "type": "[]int32", "default": "null", "entityType": "brokerId",
+      "versions": "0+", "nullableVersions": "0+", "taggedVersions": "0+", "tag": 2,
+      "about": "null if the replicas didn't change; the new replicas otherwise." },
+    { "name": "RemovingReplicas", "type": "[]int32", "default": "null", "entityType": "brokerId",
```
Why is RemovingReplicas nullable here, but not in PartitionRecord.json?

cmccabe replied:
null means that the PartitionChangeRecord didn't change it. That possibility doesn't exist in PartitionRecord, which must spell out the full values of all fields.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/QuorumController.java:
```diff
@@ -1111,19 +1111,19 @@ private QuorumController(LogContext logContext,
             listPartitionReassignments(ListPartitionReassignmentsRequestData request) {
+        if (request.topics() != null && request.topics().isEmpty()) {
+            return CompletableFuture.completedFuture(
+                new ListPartitionReassignmentsResponseData().setErrorMessage(null));
```
Is there a need for setErrorMessage(null)? It seem the error message defaults to null.

cmccabe replied:
Unfortunately, the default is not null. I could change the default, but I'd have to change a bunch of other code, so I left it as-is.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/QuorumControllerMetrics.java:
```diff
@@ -90,6 +97,13 @@ public Integer value() {
                 return preferredReplicaImbalanceCount;
             }
         });
+        this.reassigningPartitionCountGauge = omitBrokerMetrics ? null :
```
So, in the combined mode, we won't report REASSIGNING_PARTITION_COUNT? It seems it's useful in the combined mode too.

cmccabe replied:
I took this out. (We'll do metrics later...)

junrao reviewed (commented):
@cmccabe : Thanks for the PR. Made a pass of non-testing files. A few comments below.

cmccabe committed:
KAFKA-12803: Support reassigning partitions when in KRaft mode #10753

cmccabe commented:
I just finished rebasing this on the changes of the past few weeks. I removed the changes to add a new metric since the PR was getting big. I'll work on addressing all comments in a little bit, but for now here is the rebased PR...

junrao commented on metadata/src/main/java/org/apache/kafka/controller/PartitionChangeBuilder.java:
```diff
@@ -0,0 +1,229 @@
+import static org.apache.kafka.metadata.LeaderConstants.NO_LEADER_CHANGE;
+
+/**
+ * PartitionMutator handles changing partition registrations.
```
The class is no longer called PartitionMutator.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/PartitionChangeBuilder.java:
```diff
@@ -0,0 +1,229 @@
+        return this;
+    }
+
+    public PartitionChangeBuilder setAlwaysElectPreferredIfPossible(boolean alwaysElectPreferredIfPossible) {
```
Do we need this since it's only used in tests?

cmccabe replied:
Good catch. This is supposed to be set when we're handling the `electLeaders` RPC, since in that case you want to move to the preferred leader if it's not on that leader already. Should be fixed now.

cmccabe replied:
I also added a unit test for `electLeaders`

cmccabe committed:
Merge branch 'trunk' into KAFKA-12803

junrao commented on metadata/src/main/java/org/apache/kafka/controller/PartitionReassignmentRevert.java:
```diff
@@ -0,0 +1,89 @@
+                // not possible. The reason it is not possible is that doing so would
+                // create an empty ISR, which is not allowed.
+                //
+                // In this case, we leave in one of the adding replicas permanently.
```
This is different from how the old controller does. The old controller simply reverts the replicas to the original ones in KafkaController.maybeBuildReassignment(). We can change the behavior, but we probably need a KIP and do this consistently between the old and the new controller.

cmccabe replied:
It's not possible to revert to the original replicas in all cases, though. Specifically, if none of the original replicas are in the ISR, you can't do such a revert without data loss. If the old controller is doing that unconditionally, it seems like a bug.

I suppose we could also fail the revert operation, but that doesn't seem like it would be a useful behavior from the user's point of view....

junrao replied:
One weird side effect of doing this is that it can change the replication factor for a partition to be different from the original one.

cmccabe replied:
I guess we can just not allow the reassignment to be cancelled in this situation. It should still be possible to undo the reassignment by creating a new reassignment, so it doesn't totally trap the user, I suppose...

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -558,51 +608,111 @@ BrokersToIsrs brokersToIsrs() {
+                            setPartitionIndex(partitionData.partitionIndex()).
+                            setErrorCode(FENCED_LEADER_EPOCH.code()));
+                        continue;
+                    }
```
It would be useful to add an info level logging when an AlterIsr triggers the completion of the partition reassignment.

junrao commented on metadata/src/main/java/org/apache/kafka/metadata/PartitionRegistration.java:
```diff
@@ -180,6 +187,44 @@ public LeaderAndIsrPartitionState toLeaderAndIsrPartitionState(TopicPartition tp
+     * Returns true if this partition is reassigning.
+     */
+    public boolean isReassigning() {
+        return removingReplicas.length > 0 | addingReplicas.length > 0;
```
Shoud | be || ?

junrao commented on metadata/src/main/java/org/apache/kafka/metadata/PartitionRegistration.java:
```diff
@@ -180,6 +187,44 @@ public LeaderAndIsrPartitionState toLeaderAndIsrPartitionState(TopicPartition tp
+     * @param newIsr    The new ISR.
+     * @return          True if the reassignment is complete.
+     */
+    public boolean isrChangeCompletesReassignment(int[] newIsr) {
```
It seems that this is only used in tests and is duplicating the logic in PartitionChangeBuidler.completeReassignmentIfNeeded(). Do we still need it? If so, could we move this to a test util?

cmccabe replied:
Yes, this is unused now. I'll remove it

junrao commented on metadata/src/main/java/org/apache/kafka/image/TopicDelta.java:
```diff
@@ -103,7 +103,8 @@ public TopicImage apply() {
                 PartitionRegistration prevPartition = image.partitions().get(entry.getKey());
-                if (prevPartition == null || prevPartition.leader != brokerId) {
+                if (prevPartition == null ||
+                        prevPartition.leaderEpoch != entry.getValue().leaderEpoch) {
```
Should we check for partitionEpoch change too? Ditto in newLocalFollowers().

cmccabe replied:
Good point. Thinking about it more, we can just check for partition epoch here, since LE can't change without PE changing.

junrao commented on metadata/src/test/java/org/apache/kafka/controller/PartitionChangeBuilderTest.java:
```diff
@@ -0,0 +1,246 @@
+    }
+
+    @Test
+    public void testQuickReassignment() {
```
testQuickReassignment =>testRemovingReplicaReassignment ?

junrao commented on metadata/src/test/java/org/apache/kafka/controller/PartitionChangeBuilderTest.java:
```diff
@@ -0,0 +1,246 @@
+    }
+
+    @Test
+    public void testStartLongReassignment() {
```
testStartLongReassignment =>testAddingReplicaReassignment ?

junrao commented on metadata/src/test/java/org/apache/kafka/controller/PartitionChangeBuilderTest.java:
```diff
@@ -0,0 +1,246 @@
+        assertEquals(Optional.of(new ApiMessageAndVersion(new PartitionChangeRecord().
+                setTopicId(FOO_ID).
+                setPartitionId(0).
+                setReplicas(Arrays.asList(1, 2)).
```
In the removal case, the replicas should include the removed replica, right?

cmccabe replied:
This reassignment completes immediately because it's a pure removal (nothing added) and the ISR is ready to go immediately.

(It would be more typical to also add a new node to replace the one being removed, but in this case we aren't)

junrao commented on metadata/src/test/java/org/apache/kafka/metadata/PartitionRegistrationTest.java:
```diff
@@ -105,4 +105,42 @@ public void testToLeaderAndIsrPartitionState() {
+    @Test
+    public void testPartitionControlInfoIsrChangeCompletesReassignment() {
+        PartitionRegistration partition0 = new PartitionRegistration(
+            new int[]{1, 2, 3, 4}, new int[]{3}, new int[]{3}, new int[] {}, 1, 0, 0);
```
Could new int[] {} be Replicas.NONE?

cmccabe replied:
Good idea. But looking again, we can remove this test because the function has been shifted out of PartitionRegistration

junrao reviewed (commented):
@cmccabe : Thanks for the updated PR. A few more comments.

junrao commented on core/src/test/scala/integration/kafka/server/RaftClusterTest.scala:
```diff
@@ -422,4 +351,75 @@ class RaftClusterTest {
+        reassignments.put(new TopicPartition("foo", 0),
+          Optional.of(new NewPartitionReassignment(Arrays.asList(2, 1, 0))))
+        reassignments.put(new TopicPartition("foo", 1),
+          Optional.of(new NewPartitionReassignment(Arrays.asList(0, 1, 2))))
```
Could we do an expansion test?

cmccabe replied:
Good idea

junrao commented on metadata/src/test/java/org/apache/kafka/metadata/ReplicasTest.java:
```diff
@@ -88,9 +90,34 @@ public void testCopyWithout() {
+    @Test
+    public void testToSet() {
+        assertEquals(Collections.emptySet(), Replicas.toSet(new int[] {}));
+        assertEquals(new HashSet<>(Arrays.asList(3, 1, 5)),
```
Could we add a test where the input has duplicates?

junrao commented on metadata/src/test/java/org/apache/kafka/controller/ReplicationControlManagerTest.java:
```diff
@@ -763,17 +747,227 @@ public void testValidateBadManualPartitionAssignments() throws Exception {
+                        setErrorCode(NO_REASSIGNMENT_IN_PROGRESS.code()).
+                        setErrorMessage(null)))))),
+            cancelResult);
+        System.out.println("running final alterIsr...");
```
Hmm, should we do logging instead of printing to System.out? Ditto below.

junrao commented on metadata/src/test/java/org/apache/kafka/controller/ReplicationControlManagerTest.java:
```diff
@@ -763,17 +747,227 @@ public void testValidateBadManualPartitionAssignments() throws Exception {
+                    new ReassignablePartitionResponse().setPartitionIndex(0).
+                        setErrorMessage(null)))))),
+            cancelResult);
+        ControllerResult<AlterIsrResponseData> alterIsrResult = replication.alterIsr(
```
Is the alterIsr meant to be tested before cancellation? If so, probably move the code up to make it clear.

cmccabe committed:
Merge branch 'trunk' into KAFKA-12803

cmccabe committed:
Address some review comments

- Fix PartitionChangeBuilder JavaDoc

- ReplicationControlManager# electLeaders must set alwaysElectPreferredIfPossible

- ReplicationControlManagerTest remove printlns

cmccabe committed:
Merge branch 'trunk' into KAFKA-12803

cmccabe committed:
Address more review comments

cmccabe committed:
TopicDelta: leaderEpoch -> partitionEpoch in comment

junrao commented on metadata/src/main/java/org/apache/kafka/controller/PartitionChangeBuilder.java:
```diff
@@ -0,0 +1,243 @@
+        final boolean unclean;
+
+        BestLeader() {
+            for (int replica : targetIsr) {
```
Hmm, replicas in targetIsr is not necessarily ordered in the same way as the order in the replica assignment. To favor preferred leader, it seems that we need to iterate the replicas in the order in targetReplicas?

cmccabe replied:
Good point

junrao commented on metadata/src/main/java/org/apache/kafka/controller/PartitionReassignmentRevert.java:
```diff
@@ -0,0 +1,105 @@
+                throw new InvalidReplicaAssignmentException("Invalid replica " +
+                    "assignment: addingReplicas contains all replicas.");
+            }
+            isr.add(replicas.get(0));
```
Hmm, do we need to change isr here? It seems that BestLeader handles the unclean leader election with empty isr already.

cmccabe replied:
It seems simpler to do it here because then we can reject the unclean case before creating the `PartitionChangeBuilder`

junrao replied:
It seems that setting the unclean flag is enough to bypass PartitionChangeBuilder?

cmccabe replied:
So the intention is that if the revert is "unclean" and we don't have unclean leader election enabled, the revert operation for the partition returns an error rather than succeeding.

cmccabe committed:
Reposition cancellation code in ReplicationControlManagerTest a bit to make it clearer

junrao reviewed (commented):
@cmccabe : Thanks for the updated PR. A couple of more comments below.

cmccabe committed:
PartitionChangeBuilder: make sure we select replicas in the order they appear in replicas list

junrao reviewed (approved):
@cmccabe : Thanks for the updated PR. LGTM. Just a minor comment below.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/PartitionChangeBuilder.java:
```diff
@@ -0,0 +1,243 @@
+     *
+     * Changes that do NOT fall in these categories will increase the partition epoch, but
+     * not the leader epoch. Note that if the leader epoch increases, the partition epoch
+     * will always increases as well; there is no case where the partition epoch increases
```
increases => increase

cmccabe committed:
Fix comment a bit

cmccabe committed:
Merge branch 'trunk' into KAFKA-12803

jacky1193610322 commented:
hello, I don't see the logic delete the log of removing replica, now the code only manager the meta, not call stop replica as before.

dengziming commented:
@jacky1193610322 , The metadata will write to KafkaRaftClient, and the raft listener will handle this. see `QuorumMetaLogListener`.

jacky1193610322 commented:
@dengziming thank you, I have seen the QuorumMetaLogListener and BrokerMetadataListener class before. In this pr, it may be set targetRemoving field when reassigning partitions, and the broker should delete the log of removing replica. but i don't see the logic.

dengziming commented:
@jacky1193610322 , After `BrokerMetadataListener` get the partition change message, the `MetadataPublisher` will publish change `replicaManager` who will remove the related replicas, you can refer to `BrokerMetadataPublisher` to get the details.

jacky1193610322 commented:
@dengziming Thank you very much, but calculateDeltaChanges method only contain the related topic of RemoveTopicRecord msg, the removing replicas of PartitionChangeRecord is not handled.

junrao commented:
Thanks for the comment @jacky1193610322. Good point. 

@cmccabe : Could you check if this is an issue? It seems that ReplicaManager.calculateDeltaChanges() doesn't take PartitionChangeRecord into consideration to determine the removed replicas?
